File: Dome/dome/rules/sqli.py
"""SQL Injection detection rules."""
import re

# Patterns ordered from high-confidence to heuristic
PATTERNS = [
    # Classic UNION-based
    (r"(?i)\bunion\b.{0,20}\bselect\b",                          "SQLI-001", "UNION SELECT injection"),
    # Boolean-based blind
    (r"(?i)\b(and|or)\b\s+[\w'\"]+\s*[=<>!]+\s*[\w'\"]+",       "SQLI-002", "Boolean-based blind SQLi"),
    # Stacked queries / comment sequences
    (r"(--|#|/\*.*?\*/)\s*$",                                     "SQLI-003", "SQL comment terminator"),
    # Sleep / time-based blind
    (r"(?i)\b(sleep|benchmark|waitfor\s+delay|pg_sleep)\s*\(",   "SQLI-004", "Time-based blind SQLi"),
    # INFORMATION_SCHEMA probing
    (r"(?i)\binformation_schema\b",                               "SQLI-005", "Schema enumeration"),
    # Common SQLi payloads
    (r"(?i)'?\s*(OR|AND)\s+'?1'?\s*=\s*'?1",                     "SQLI-006", "Tautology (1=1)"),
    # Quote + keyword
    (r"(?i)['\"];\s*(DROP|ALTER|INSERT|UPDATE|DELETE|EXEC)\b",   "SQLI-007", "Destructive statement after quote"),
    # xp_cmdshell / sp_executesql
    (r"(?i)\b(xp_cmdshell|sp_executesql|exec\s*\()",             "SQLI-008", "MSSQL stored procedure abuse"),
    # Load/outfile
    (r"(?i)\b(load_file|into\s+outfile|into\s+dumpfile)\b",      "SQLI-009", "File read/write via SQL"),
    # Hex encoding bypass
    (r"(?i)0x[0-9a-f]{4,}",                                      "SQLI-010", "Hex-encoded payload"),
]

COMPILED = [(re.compile(p), rid, desc) for p, rid, desc in PATTERNS]


def check(value: str) -> list[dict]:
    hits = []
    for pattern, rule_id, description in COMPILED:
        if pattern.search(value):
            hits.append({"rule_id": rule_id, "description": description, "category": "sqli"})
    return hits

File: Dome/dome-rs/src/engine/sqli.rs
/// SQL injection detection – operates on already-normalised input.
use once_cell::sync::Lazy;
use regex::Regex;
use super::Hit;

const SCORE: u32 = 25;

static RULES: Lazy<Vec<(Regex, &'static str, &'static str)>> = Lazy::new(|| {
    vec![
        (Regex::new(r"union\s.{0,30}select").unwrap(),
         "SQLI-001", "UNION SELECT injection"),
        (Regex::new(r"(and|or)\s+[\w']+\s*[=<>!]+\s*[\w']+").unwrap(),
         "SQLI-002", "Boolean tautology"),
        (Regex::new(r"(--|#)\s*$").unwrap(),
         "SQLI-003", "SQL comment terminator"),
        (Regex::new(r"\b(sleep|benchmark|waitfor\s+delay|pg_sleep)\s*\(").unwrap(),
         "SQLI-004", "Time-based blind injection"),
        (Regex::new(r"\binformation_schema\b").unwrap(),
         "SQLI-005", "Information schema enumeration"),
        (Regex::new(r"'?\s*(or|and)\s+'?1'?\s*=\s*'?1").unwrap(),
         "SQLI-006", "Classic tautology or 1=1"),
        (Regex::new(r"[';]\s*(drop|alter|insert|update|delete|exec)\b").unwrap(),
         "SQLI-007", "Destructive statement after quote"),
        (Regex::new(r"\b(xp_cmdshell|sp_executesql)\b").unwrap(),
         "SQLI-008", "MSSQL stored procedure abuse"),
        (Regex::new(r"\b(load_file|into\s+outfile|into\s+dumpfile)\b").unwrap(),
         "SQLI-009", "SQL file read/write"),
        (Regex::new(r"0x[0-9a-f]{4,}").unwrap(),
         "SQLI-010", "Hex-encoded payload"),
        (Regex::new(r";\s*(select|insert|update|delete|drop|exec)\b").unwrap(),
         "SQLI-011", "Stacked queries"),
        (Regex::new(r"\bcase\s+when\b.{0,60}\bthen\b").unwrap(),
         "SQLI-012", "CASE WHEN blind injection"),
        (Regex::new(r"\border\s+by\s+\d+").unwrap(),
         "SQLI-013", "ORDER BY column index probe"),
        (Regex::new(r"(@@version|version\s*\(\s*\)|database\s*\(\s*\)|user\s*\(\s*\))").unwrap(),
         "SQLI-014", "Database fingerprinting"),
    ]
});

pub fn check(normalised: &str) -> Vec<Hit> {
    RULES.iter().filter_map(|(re, id, desc)| {
        if re.is_match(normalised) {
            Some(Hit { rule_id: id, description: desc, category: "sqli", score: SCORE })
        } else {
            None
        }
    }).collect()
}
